tests/e2e/produce_consume.go

```diff
@@ -91,3 +93,89 @@ func requireInternalProducingConsumingMessage(kubectlOptions k8s.KubectlOptions,
 		Expect(consumedMessages).Should(ContainSubstring(currentTime.String()))
 	})
 }
+
+// requireExternalProducingConsumingMessage gets the Kafka cluster external addresses from the kafkaCluster CR status
+// when externalAddresses is not specified. It also produces and consumes messages and makes a comparison between them.
+func requireExternalProducingConsumingMessage(kubectlOptions k8s.KubectlOptions, topicName, tlsSecretName string, externalAddresses ...string) {
+	It("Producing and consuming messages", func() {
+		if len(externalAddresses) == 0 {
+			var err error
+			externalAddresses, err = getExternalListenerAddresses(kubectlOptions, "", kafkaClusterName)
+			Expect(err).ShouldNot(HaveOccurred())
+		}
+
+		var tlsConfig *tls.Config
+		var clientOptions []kgo.Opt
+		if tlsSecretName != "" {
+			var err error
+			tlsConfig, err = getTLSConfigFromSecret(kubectlOptions, tlsSecretName)
+			Expect(err).ShouldNot(HaveOccurred())
+			clientOptions = append(clientOptions, kgo.DialTLSConfig(tlsConfig))
+		}
+
+		message := time.Now().String()
+
+		err := producingMessagesExternally(externalAddresses, topicName, []string{message}, clientOptions...)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		consumedMessages, err := consumingMessagesExternally(externalAddresses, topicName, clientOptions...)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		By(fmt.Sprintf("Comparing produced: '%s' and consumed message: '%s'", message, consumedMessages))
+		found := false
+		for i := range consumedMessages {
+			if consumedMessages[i] == message {
+				found = true
+				break
+			}
+		}
+		Expect(found).Should(BeTrue())
+	})
+}
+
+// getExternalListenerAddresses gets the Kafka cluster external addresses from the kafkaCluster CR.
+// When externalListenerName is not specified it uses the first externalListener name in the CR to get addresses.
+func getExternalListenerAddresses(kubectlOptions k8s.KubectlOptions, externalListenerName, kafkaClusterName string) ([]string, error) {
+	By(fmt.Sprintf("Getting Kafka cluster '%s' external listener addresses", kafkaClusterName))
+	if externalListenerName == "" {
+		By("Getting external listener names")
+		externalListenerNames, err := getK8sResources(kubectlOptions,
+			[]string{kafkaKind},
+			"",
+			kafkaClusterName,
+			kubectlArgGoTemplateExternalListenersName,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("getting external listeners name: %w", err)
+		}
+		if len(externalListenerNames) == 0 {
+			return nil, fmt.Errorf("external listener %w", ErrorNotFound)
+		}
```

Comment:
**Note**: This could be a function on its own to list external listener names, sounds like something we might want to reuse in the future, it's a refactor for later.

Reply:
I agree. Im going to refactor it into separate function